Random VS/Mac compile failures for sharing violation

MacOS Big Sur 11.1
Visual Studio for Mac 8.8.5 (build 18)
Avalonia 0.9.12

About a quarter of the time after correcting a syntax error that caused a build failure, the next build attempt fails with the following message:

/Users/t/.nuget/packages/avalonia/0.9.12/build/AvaloniaBuildTasks.targets(5,5): Error MSB4018: The "CompileAvaloniaXamlTask" task failed unexpectedly.
System.IO.IOException: Sharing violation on path /Users/t/Projects/PhotoViewer/obj/Debug/netcoreapp5.0/osx-x64/Avalonia/original.pdb
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00259] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:274 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:106 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.FileSystem.CopyFile (System.String sourceFullPath, System.String destFullPath, System.Boolean overwrite) [0x00035] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:55 
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName, System.Boolean overwrite) [0x00056] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/File.cs:74 
  at Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute () [0x00074] in D:\a\1\s\src\Avalonia.Build.Tasks\CompileAvaloniaXamlTask.cs:29 
  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00023] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:577 
  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002b9] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:809  (MSB4018) (PhotoViewer)

The only workaround I know of is to close and reopen the project, after which it compiles successfully (if there are no other errors),

Hi!

We have faced with this issue too. To be honest every 3rd compilation pipeline fails on our compilation servers so the working process has stopped with endless reruns. We are in fire with this bug...

```
C:\Users\<user>\.nuget\packages\avalonia\11.0.4\buildTransitive\AvaloniaBuildTasks.targets(120,5): error MSB4018: непредвиденная ошибка при выполнении задачи "CompileAvaloniaXamlTask".
System.IO.IOException: Процесс не может получить доступ к файлу "obj\.sual\Debug\net7.0\Avalonia\original.pdb", так как этот файл используется другим процессом.
   в System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   в System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   в Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() в /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:строка 36
   в Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   в Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [<here is the path to failed csproj>]
```

@kekekeks @maxkatz6  or other maintainers: here is almost 100% repro , the root cause is the multilply `root` projects, all of them is referencing (via others csprojes) to `BBBBBB` project, and `BBBBBB` project fails with 

```
C:\Users\<username>\.nuget\packages\avalonia\11.0.4\buildTransitive\AvaloniaBuildTasks.targets(120,5): error MSB4018: непредвиденная ошибка при выполнении задачи "CompileAvaloniaXamlTask".
System.IO.IOException: Процесс не может получить доступ к файлу "obj\Debug\net7.0\Avalonia\original.pdb", так как этот файл используется другим процессом.
   в System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   в System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   в Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() в /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:строка 36
   в Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   в Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
```

[master_A2Z.zip](https://github.com/AvaloniaUI/Avalonia/files/13197489/master_A2Z.zip)


Interestingly, compilation inside VS is completely fine, so one needs to compile from command line to see the bug. Here is  repro powershell script:

```
Set-Location 'path to master_A2Z folder';

$RestoreProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'dotnet' -ArgumentList ( ' restore A2Z.sln ' );
Wait-Process -InputObject $RestoreProcess;

$Iteration = 0;
while(1)
{
    [System.Console]::WriteLine('Iteration: ' + $Iteration + ', Now: ' + [System.DateTime]::Now);

    # probably you will need to change this msbuild path
    $MsBuildProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\MSBuild.exe' -ArgumentList ( ' A2Z.sln /v:quiet /bl:mk.binlog /maxcpucount:1 /p:BuildInParallel=false' );
    Wait-Process -InputObject $MsBuildProcess;
    if($MsBuildProcess.ExitCode -ne 0)
    {
        [System.Console]::WriteLine('msbuild fail!');
        exit $MsBuildProcess.ExitCode;
    }

    $Iteration += 1;
}
```

running this, I see the following:

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/29/2023 12:33:42
msbuild fail!
```

or rarely this:

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/29/2023 12:26:08
Iteration: 1, Now: 10/29/2023 12:32:07
msbuild fail!
```


looking at `MSBuild Structured Log Viewer` I see this error.

here is Process Monitor log

```
13:01:18,3088972	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Read/Write, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3092621	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3096137	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3099288	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3102536	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3105866	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3109093	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3113714	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3118344	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:23,1389123	MSBuild.exe	142976	IRP_MJ_CLEANUP	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
13:01:23,1389723	MSBuild.exe	142976	IRP_MJ_CLOSE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
13:01:23,4386530	System	4	FASTIO_ACQUIRE_FOR_SECTION_SYNCHRONIZATION	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	SyncType: SyncTypeOther
13:01:23,4387064	System	4	IRP_MJ_CLOSE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
```

if I read this correctly, MSBuild conflicts with itself (PTAL on `/maxcpucount:1 /p:BuildInParallel=false` in powershell script).


I hope any maintainer will take a look here, because this is very difficult problem for those who faced it. We are going to perform autorerun if  building fails and `original.pdb` contains in build log, but our compilation cluster will degrade severely because of this.

**Please let me know if you will need any info\support.**

--- a/src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs
+++ b/src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs
@@ -63,8 +63,8 @@ namespace Avalonia.Build.Tasks
             try
             {
                 references = references.Where(r => !r.ToLowerInvariant().EndsWith("avalonia.build.tasks.dll")).ToArray();
-                var typeSystem = new CecilTypeSystem(references, input);
-                var refTypeSystem = !string.IsNullOrWhiteSpace(refInput) && File.Exists(refInput) ? new CecilTypeSystem(references, refInput) : null;
+                using var typeSystem = new CecilTypeSystem(references, input);
+                using var refTypeSystem = !string.IsNullOrWhiteSpace(refInput) && File.Exists(refInput) ? new CecilTypeSystem(references, refInput) : null;
 
                 var asm = typeSystem.TargetAssemblyDefinition;
                 var refAsm = refTypeSystem?.TargetAssemblyDefinition;

